scripts: Create CVE assistance script

d3zd3z opened the pull request:
This script is primarily intended to be used by the security team to
help generate release notes and CVE database entries based on the
security JIRA instance.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
scripts: Create CVE assistance script

This script is primarily intended to be used by the security team to
help generate release notes and CVE database entries based on the
security JIRA instance.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
scripts: security: Skip rejected subtasks

Rejected subtasks mean that particular branch is not affected by the
issue.  Avoid trying to print information about that branch.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
scripts: security: Make rnote link targets unique

Restructured text requires that the targets of links be unique.  Add
information about the CVE/issue/pr number to the link text so that they
will all be unique.

This will still produce duplicates, however, if there are multiple CVEs
that use the same PR.  These will need to be fixed by hand.

Signed-off-by: David Brown <david.brown@linaro.org>

ceolin commented on scripts/security/cve-template.json:
Is this a valid attribute ? I didn't see it.

d3zd3z replied:
This json is a direct copy of one generated by the vulnogram tool.

ceolin replied:
will look it, looking the scheme I didn't see this property

ceolin commented on scripts/security/cve-template.json:
Since it is a template, we shouldn't have info about an specific issue.

d3zd3z replied:
True. It gets replaced entirely.

ceolin commented on scripts/security/cve-template.json:
ditto

ceolin commented on scripts/security/tocve.py:
```diff
@@ -0,0 +1,225 @@
+        for link in issue.get_full_links():
+            print("- `PR fix for {}".format(link["ver"]))
+            print("  <{}>`_".format(link["URL"]))
+            print()
```
Would be clear if you use a string template to do this

d3zd3z replied:
You mean string interpolation, or f-Strings? String templates are also a python thing, but are more complex. I can use f-Strings, I just keep forgetting they are even there.

ceolin replied:
f-strings would make it simpler

ceolin commented on scripts/security/tocve.py:
```diff
@@ -0,0 +1,225 @@
+        self.links = query_links(issue)
+        # print("ticket: ", self.ticket)
+        # print("links: ", self.links)
+        # print("done")
```
don't forget remote these prints for debug

d3zd3z replied:
Do they hurt anything? Well, at least the ones that can print out the ticket or links. The 'done' not so useful.

ceolin replied:
we don't have these leftovers in other scripts

d3zd3z committed:
scripts: security: Remove details from CVE template

Remove specific information from the CVE template, indicating things
that must either be filled in by the script, or edited by the user.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
scripts: security: Cleanups in cve script

Fix some of the formatting under various types of configurations of CVE
data.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
scripts: security: Add license to tocve script

Add a SPDX header to the script to clarity the license.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
CODEOWNERS: Add security scripts

Set myself (@d3zd3z) as the codeowner for the security scripts.  These
scripts are used by the security team to process and manage
vulnerability reports.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z commented:
Once we agree on the content, I can squash this into one or two commits (one for the script/template, and one for the CODEOWNERS update).

d3zd3z committed:
scripts: security: Add general scan.py script

This is the start of a script to scan JIRA and github to correlate
changes in PR status with JIRA security issues.

This expects two tokens for authentication.  The jira auth token should
be placed in ~/.netrc as follows:

   machine zephyrprojectsec.atlassian.net
     login <email>
     password <token>

The token for github should also be placed in this file:

   machine github.com
     login <github-id>
     passwrd <github-token>

Signed-off-by: David Brown <david.brown@linaro.org>

ceolin commented on scripts/security/scan.py:
```diff
@@ -0,0 +1,139 @@
+repo = gh.get_repo("zephyrproject-rtos/zephyr")
+gitwork = Git(".")
+
+with open('token') as fd:
```
This block is not used. Also it is trying to access this local file that is not documented.

ceolin commented on scripts/security/scan.py:
```diff
@@ -0,0 +1,139 @@
+#! /usr/bin/env python3
+
+from github import Github
+from git import Git
```
These dependencies have to be added to `requirements.txt`. Maybe an exclusive file for security. `requirements-security.txt`

ceolin commented on scripts/security/scan.py:
```diff
@@ -0,0 +1,139 @@
+                pr = int(m.group(1))
+                gpr = repo.get_pull(pr)
+                print("        ", pr, gpr.state, gpr.title)
+                for c in gpr.get_commits():
```
Why you iterating over all commits in the pr ? This won't work because the hash changes when rebased. So, if you want to check if it was merged we can just check it in the pr doing `gpr.merged`. If you want to know which branch it was merged, then we need to get the merge commit hash:
```
sha = gpr.merge_commit_sha
gitwork.describe(sha)
```

But we don't need to iterate over all commits. If you really need/want to to that I think you need to iterate the number of commits starting from `gpr.merge_commit_sha`, but I'm not seeing why we'd need that.

ceolin commented on scripts/security/scan.py:
```diff
@@ -0,0 +1,139 @@
+        if isinstance(j, list):
+            return j
+        # for k in j.keys():
+        #     print(k)
```
just remove these commented code :)

d3zd3z commented:
Closing this. These scripts have been moved to their own repo: https://github.com/zephyrproject-rtos/swg-scripts/
